flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapRestoreOperation.java

```diff
@@ -143,6 +143,7 @@ public Void restore() throws Exception {
 					}
 
 					keySerializerRestored = true;
+					keyContext.setCurrentKeySerializer(keySerializerProvider.currentSchemaSerializer());
```

Comment:
Because the serializer might be re-configured in `setPreviousSerializerSnapshotForRestoredState`, please check the [implementation](https://github.com/apache/flink/blob/3d5b6b166202c423b1529316bb9a9a755628ae1a/flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSerializerProvider.java#L335) in `StateSerializerProvider.EagerlyRegisteredStateSerializerProvider`

Reply:
I have the feeling that the 2 separate places to call `setCurrentKeySerializer` (first when it is initialized in the builder, as well as here) is going to a bit error prone in the future as the dependency is very fragile.

Maybe the way to go is actually to let the `InternalKeyContextImpl` wrap the `StateSerializerProvider`, so that the handling of the serializer reconfiguration is handled in the context already? WDYT?